internal/refactoring/remove_statement.go

```diff
@@ -0,0 +1,122 @@
+// Copyright (c) HashiCorp, Inc.
+// SPDX-License-Identifier: MPL-2.0
+
+package refactoring
+
+import (
+	"fmt"
+
+	"github.com/hashicorp/hcl/v2"
+	"github.com/opentofu/opentofu/internal/addrs"
+	"github.com/opentofu/opentofu/internal/configs"
+	"github.com/opentofu/opentofu/internal/tfdiags"
+)
+
+type RemoveStatement struct {
+	From      addrs.ConfigRemovable
+	DeclRange tfdiags.SourceRange
+}
+
+// GetEndpointsToForget recurses through the modules of the given configuration
+// and returns an array of all "removed" addresses within, in a
+// deterministic but undefined order.
+// We also validate that the removed modules/resources configuration blocks were removed.
+func GetEndpointsToForget(rootCfg *configs.Config) ([]addrs.ConfigRemovable, tfdiags.Diagnostics) {
+	rm := findRemoveStatements(rootCfg, nil)
+	diags := validateRemoveStatements(rootCfg, rm)
+	removedAddresses := make([]addrs.ConfigRemovable, len(rm))
+	for i, rs := range rm {
+		removedAddresses[i] = rs.From
+	}
+	return removedAddresses, diags
+}
+
+func findRemoveStatements(cfg *configs.Config, into []*RemoveStatement) []*RemoveStatement {
+	modAddr := cfg.Path
+
+	for _, rc := range cfg.Module.Removed {
+		var removedEndpoint *RemoveStatement
+		switch FromAddress := rc.From.RelSubject.(type) {
+		case addrs.ConfigResource:
+			// Get the absolute address of the resource by appending the module config address
+			// to the resource's relative address
+			absModule := make(addrs.Module, 0, len(modAddr)+len(FromAddress.Module))
+			absModule = append(absModule, modAddr...)
+			absModule = append(absModule, FromAddress.Module...)
+
+			var absConfigResource addrs.ConfigRemovable = addrs.ConfigResource{
+				Resource: FromAddress.Resource,
+				Module:   absModule,
+			}
+
+			removedEndpoint = &RemoveStatement{From: absConfigResource, DeclRange: tfdiags.SourceRangeFromHCL(rc.DeclRange)}
+
+		case addrs.Module:
+			// Get the absolute address of the module by appending the module config address
+			// to the module itself
+			var absModule = make(addrs.Module, 0, len(modAddr)+len(FromAddress))
+			absModule = append(absModule, modAddr...)
+			absModule = append(absModule, FromAddress...)
+			removedEndpoint = &RemoveStatement{From: absModule, DeclRange: tfdiags.SourceRangeFromHCL(rc.DeclRange)}
+
+		default:
+			panic(fmt.Sprintf("unhandled address type %T", FromAddress))
```

Comment:
Would this be hit if a data block was referenced or is that handled by ConfigResource?

Reply:
It'll not. However, I did miss the validation for the `data` block. And I added it [here](https://github.com/opentofu/opentofu/pull/1158/commits#:~:text=if%20removed%20target%20is%20data%20source%2C%20throw%20an%20error)